File: JS/백준 알고리즘/1774.js
input = require("fs").readFileSync("예제.txt").toString().trim().split("\n");

// variable
let point = [];
let grahp = [];
let parent = [];
let rank = [];
let answer = 0;

// Helper function
function getDist(point1, point2) {
  X = Math.abs(point1[0] - point2[0]);
  Y = Math.abs(point1[1] - point2[1]);
  return Math.sqrt(X * X + Y * Y);
}

// kruskal Algorithm : find function
function find(node) {
  if (parent[node] != node) {
    parent[node] = find(parent[node]);
  }
  return parent[node];
}
// kruskal Algorithm : union function
function union(node1, node2) {
  root1 = find(node1);
  root2 = find(node2);

  if (rank[root1] > rank[root2]) {
    parent[root2] = root1;
  } else {
    parent[root1] = root2;
    if (rank[root1] == rank[root2]) rank[root2] += 1;
  }
}

// kruskal Algorithm : makeSet function
function makeSet(node) {
  parent[node] = node;
  rank[node] = 0;
}
// Main function
function solution(input) {
  // 1st line
  [godNum, already] = input.shift().split(" ").map(Number);

  // 2nd line
  for (let i = 0; i < godNum; i++) {
    [x, y] = input[i].split(" ").map(Number);
    point.push([x, y]);

    // make node
    makeSet(i);
  }

  // make grahp
  for (let i = 0; i < point.length; i++) {
    for (let j = i + 1; j < point.length; j++) {
      grahp.push([getDist(point[i], point[j]), i, j]);
    }
  }

  // already union
  for (let j = godNum; j < godNum + already; j++) {
    // put in already union
    [node1, node2] = input[j].split(" ").map(Number);
    union(node1 - 1, node2 - 1);
  }

  // grahp sort
  grahp.sort((a, b) => a[0] - b[0]);
  console.log(grahp);
  grahp.forEach((el) => {
    [dist, from, to] = el;
    if (find(from) != find(to)) {
      union(from, to);
      answer += dist;
    }
  });

  console.log(answer.toFixed(2));
}
solution(input);

File: Python/백준 알고리즘/1774.py
from collections import defaultdict
import math


def getDist(point1, point2):
    X = abs(point1[0] - point2[0])
    Y = abs(point1[1] - point2[1])
    return math.sqrt((X*X) + (Y*Y))


def find(node):
    if parent[node] != node:
        parent[node] = find(parent[node])
    return parent[node]


def union(node_v, node_u):
    root1 = find(node_v)
    root2 = find(node_u)

    if rank[root1] > rank[root2]:
        parent[root2] = root1
    else:
        parent[root1] = root2
        if rank[root1] == rank[root2]:
            rank[root2] += 1


def make_set(node):
    parent[node] = node
    rank[node] = 0


god_num, alreay = list(map(int, input().split()))

parent = dict()
rank = dict()
point = list()
graph = defaultdict(list)
mst = list()
answer = 0

for _ in range(god_num):
    x, y = list(map(int, input().split()))
    point.append([x, y])

length = len(point)

for i in range(length):
    graph['vertices'].append(i)

[al_p1, al_p2] = list(map(int, input().split()))
al_p1 -= 1
al_p2 -= 1
for i in range(god_num):
    for j in range(god_num):
        graph['edges'].append([getDist(point[i], point[j]), i, j])
edges = graph['edges']


for node in graph['vertices']:
    make_set(node)

union(al_p1, al_p2)
edges.sort()
for edge in edges:
    weight, node_v, node_u = edge
    if find(node_v) != find(node_u):
        union(node_v, node_u)
        mst.append(edge)
        answer += edge[0]
print('%.2f' % answer)
